### DISCUSSION TITLE
Joinfaces Maven/Gradle Plugins vs Spring AOT

### DISCUSSION THREAD

**Discussion by larsgrefer** [joinfaces/joinfaces#1886] [Ideas]
During Spring Boot 2.x times, we implemented a Maven plugin and a Gradle plugin to perform some classpath scanning at build-time in order to save some time during application startup.
A proposal to add this to Spring Boot itself was declined at that time: https://github.com/spring-projects/spring-boot/issues/17739

With Spring Boot 3.x, there is now a pre-processing step directly supported by Spring Boot which they call "Ahead-of-Time Processing". Spring Boot implemented this in order to enable building native images with GraalVM, but it can also be used to improve the performance of typical JVM deployments: https://docs.spring.io/spring-boot/docs/3.1.4/reference/htmlsingle/#deployment.efficient.aot

In order to support GraalVM with JoinFaces it makes sense to integrate our classpath scanning with Springs AOT processing.
This would leave us with multiple classpath scanning implementations (at runtime, our Gradle/Maven plugins, and Spring AOT)

We could use this opportunity to move all our build-time processing to Spring AOT and remove our own Gradle/Maven plugins entirely. Users of our plugins would then have to migrate their builds to Spring AOT as described here: https://docs.spring.io/spring-boot/docs/3.1.4/reference/htmlsingle/#deployment.efficient.aot

@persapiens What do you think?

**Comment by persapiens** [comment]
Hi, @larsgrefer , nice to meet you.

I like the idea of integrating our classpath scanning with Springs AOT processing. I think it will enable us to work with Spring Native too.

However, removing our plugins will force the Joinfaces applications to use Spring AOT to gather our classpath performance optimizations. 

Currently, are the apps ready to work with Spring AOT? I read some explanation here: https://www.baeldung.com/spring-6-ahead-of-time-optimizations

**Reply by larsgrefer** [reply]
The Baeldung article is a good starting point, but it mixes a few points together.

Producing a proper native image with GraalVM for a Spring Boot application requires essentially two steps:

1. Preparation steps on the Spring (Boot) side
2. Actual AOT compilation by GraalVM

The first step is performed by the `processAot` Gradle task (or the `spring-boot:process-aot` Maven goal). A lot of additional configuration files and additional java code which creates all the bean definitions without component scanning are created here.

The second step then does most of the things mentioned in the article (Dead Code Elimination, Inlining, Constant Propagation, Inter-procedural optimization)

The idea behind the article and behind https://docs.spring.io/spring-boot/docs/3.1.4/reference/htmlsingle/#deployment.efficient.aot is to only do the first step, but not necessarily the second one. Thus preparing some things ahead of time at build time, but then still using a regular JVM as runtime.

Ignoring our own Gradle/Maven Plugins for a moment, the following scenarios are possible:

1. Doing everything as before (Boot 2.x): No aot whatsoever, component scanning and classpath scanning during application startup.
2. Invoking AOT preparation during build time by using the tasks/goals provided by Spring Boot, but not setting `spring.aot.enabled` at runtime and not using GraalVM.
3. Invoking AOT preparation during build time by using the tasks/goals provided by Spring Boot and setting `spring.aot.enabled` at runtime.
4. Invoking AOT preparation during build time by using the tasks/goals provided by Spring Boot and using GraalVM.

To get scenarios 3 and 4 working with JoinFaces, we need to integrate our own classpath scanning with the Spring Boot AOT processing.

After implementing this, we could then go one step further by removing our own Build tool plugins and telling users to just use scenario 2 or 3. Without JoinFaces, scenario 2 would be useless, but if we perform our classpath scanning during Spring Boots AOT processing, it could effectively replace our plugins.

**Reply by persapiens** [reply]
Thanks for the explanation. Let's do it.
